### PR TITLE
Add next_census_display to User

### PR DESCRIPTION
**Pull request by drewsamnick**
Add a next_census_display field to User. This will be used to keep track of when to show the new census teacher banner.

### CODE REVIEW

File: dashboard/app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -33,4 +33,33 @@ def post_ui_tip_dismissed
...
+    year = today.year
+    # Find the next scheduled date that is at least 2 weeks away
+    scheduled_display_dates = [1, 3, 5, 11].map {|month| Date.new(year, month, 15)}
+    scheduled_display_dates.push Date.new(year + 1, 1, 15)
+    next_date = scheduled_display_dates.select {|d| d > today && d > today + 2.weeks} [0]
+
+    @user.next_census_display = next_date
+    @user.save
+
+    render status: 200, json: {set: "next_census_display"}
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/users_controller.rb]: What is the point of this response body? Why is hardcoded `{set: "next_census_display"}` useful? Should we return the new census display time? Or no body `head 200` or `head :ok`

File: dashboard/app/controllers/api/v1/users_controller.rb [hunk 2]
```diff
@@ -33,4 +33,33 @@ def post_ui_tip_dismissed
...
+    today = Date.today
+    year = today.year
+
+    # if they dismiss Aug 1 or later then don't show until next November
+    year += 1 if today.month >= 8
+
+    @user.next_census_display = Date.new(year, 11, 15)
+    @user.save
+
+    render status: 200, json: {set: "next_census_display"}
```
Comment (aoby) [inline dashboard/app/controllers/api/v1/users_controller.rb]: same as above

File: dashboard/app/models/user.rb [hunk 3]
```diff
@@ -1603,6 +1604,12 @@ def should_see_inline_answer?(script_level)
       (script_level && UserLevel.find_by(user: self, level: script_level.level).try(:readonly_answers))
   end
 
+  def show_census_teacher_banner?
+    # Must have an NCES school to show the banner
+    users_school = try(:school_info).try(:school)
+    teacher? && users_school && (next_census_display.nil? || Date.today > next_census_display.to_date)
```
Comment (aoby) [inline dashboard/app/models/user.rb]: minor: Should this be `>=`?

File: dashboard/test/controllers/api/v1/users_controller_test.rb [hunk 4]
```diff
@@ -55,4 +55,24 @@ class Api::V1::UsersControllerTest < ActionController::TestCase
     @user.reload
     assert @user.ui_tip_dismissed_homepage_header
   end
+
+  test 'a post request to postpone_census_banner updates next_census_display' do
+    sign_in(@user)
+    @user.next_census_display = nil
+    assert @user.next_census_display.nil?
+    post :postpone_census_banner, params: {user_id: 'me'}
```
Comment (aoby) [inline dashboard/test/controllers/api/v1/users_controller_test.rb]: 1. No need to assert the thing you just set. I'd remove line 62
 2. Setting an attribute on the local `@user` object does not commit it to the DB, and will not make it available to the controller. Line 61 should be something like `@user.update!(next_census_display: nil)`, or you can create a new user object for each test case (in which case that field will already be nil).

Comment (aoby) [reply dashboard/test/controllers/api/v1/users_controller_test.rb]: Same below

File: dashboard/app/controllers/api/v1/users_controller.rb (hunk 1 above)
Comment (drewsamnick) [reply dashboard/app/controllers/api/v1/users_controller.rb]: Good point. I'm going to change it to return the new value.

File: dashboard/app/models/user.rb (hunk 3 above)
Comment (drewsamnick) [reply dashboard/app/models/user.rb]: Yes. I'll fix.

File: dashboard/test/controllers/api/v1/users_controller_test.rb (hunk 4 above)
Comment (drewsamnick) [reply dashboard/test/controllers/api/v1/users_controller_test.rb]: I had modeled this after [an existing test](https://github.com/code-dot-org/code-dot-org/blob/d14c06590bd0642663a93ef4a972578ee279661f/dashboard/test/controllers/api/v1/users_controller_test.rb#L51). Look like that wasn't a great example. I will update based on your suggestions.

Comment (aoby) [reply dashboard/test/controllers/api/v1/users_controller_test.rb]: Ick. That is a bad example. Adding a `reload` to this line https://github.com/code-dot-org/code-dot-org/blob/d14c06590bd0642663a93ef4a972578ee279661f/dashboard/test/controllers/api/v1/users_controller_test.rb#L51 will cause it to fail :(

**Comment by drewsamnick** [comment]
I updated the API to return the date that was set and cleaned up the tests.

**Pull request merged**